File: Data Structures & Algorithms/quickSort/submission-0.java
// Definition for a pair.
// class Pair {
//     int key;
//     String value;
//
//     public Pair(int key, String value) {
//         this.key = key;
//         this.value = value;
//     }
// }
class Solution {
    public List<Pair> quickSort(List<Pair> pairs) {
        quickSort(pairs, 0, pairs.size()-1);
        return pairs;
    }

    public void quickSort(List<Pair> pairs, int l, int r) {
        if(l >= r) return;
        int pivotIndex = partition(pairs,l,r);
        quickSort(pairs, l, pivotIndex - 1);
        quickSort(pairs, pivotIndex+1, r);
    }

    public int partition(List<Pair> pairs, int l, int r) {
        Pair pivot = pairs.get(r);
        int i = l-1;
        for(int j = l; j <r; j++) {
            if(pairs.get(j).key < pivot.key) {
                i++;
                swap(pairs, i, j);
            }
        }
        swap(pairs, i+1, r);
        return i+1;
    }

    public void swap(List<Pair> pairs, int a, int b) {
        Pair temp = pairs.get(a);
        pairs.set(a, pairs.get(b));
        pairs.set(b, temp);
    }
}

File: Data Structures & Algorithms/k-closest-points-to-origin/submission-0.py
class Pair:
    def __init__(self, key: int, value: int):
        self.key = key
        self.value = value

class Solution:
    def quickSort(self, dist: List[Pair], s: int, e: int, k: int) -> List[Pair]:
        if e - s + 1 <= 1:
            return dist
        
        pivot = dist[e]
        loc = s
        for i in range(s, e):
            if dist[i].value < pivot.value:
                dist[loc], dist[i] = dist[i], dist[loc]
                loc += 1
        
        dist[e] = dist[loc]
        dist[loc] = pivot
        if loc == k:
            return
        elif loc < k:
            self.quickSort(dist, loc + 1, e, k)
        else:
            self.quickSort(dist, s, loc - 1, k)
        return dist

    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        n = len(points)
        dist = [ Pair(i, points[i][0] * points[i][0] + points[i][1] * points[i][1]) for i in range(len(points))]
        self.quickSort(dist, 0, n - 1, k)
        return [points[p.key] for p in dist[:k]]
        
        

        